Add documentation to delete orphan instances functionality (#166)
kyma-project/control-plane: docs/kyma-environment-broker/03-09-subaccount-cleanup-cronjob.md

---
title: Subaccount Cleanup CronJob
type: Details
---

Each SKR instance in Kyma Environment Broker (KEB) database belongs to a global account and to a subaccount.

> Subaccount Cleanup is an application which periodically calls the CIS service and notifies about SUBACCOUNT_DELETE events. 

Reviewer: ```suggestion
Subaccount Cleanup is an application that periodically calls the CIS service and notifies about `SUBACCOUNT_DELETE` events. 
```

Based on these events, Subaccount Cleanup triggers the deprovisioning action on the SKR instance to which a given subaccount belongs.

## Details

> The Subaccount Cleanup process is divided into several steps:

Reviewer: ```suggestion
The Subaccount Cleanup workflow is divided into several steps:
```

> 1. Fetch SUBACCOUNT_DELETE events from the CIS service.

Reviewer: ```suggestion
1. Fetch `SUBACCOUNT_DELETE` events from the CIS service.
```

    a. CIS client makes a call to the CIS service and as a response, it gets a list of events divided into pages. 

    b. CIS client fetches the rest of the events by making a call to each page one by one.

>     c. A subaccount is taken from each event and kept in an array.

Reviewer: ```suggestion
    c. A subaccount ID is taken from each event and kept in an array.
```

    d. When the CIS client ends its workflow, it displays logs with information on how many subaccounts were fetched.

> 2. Find all instances in the KEB database based on the fetched subaccounts.

Reviewer: ```suggestion
2. Find all instances in the KEB database based on the fetched subaccount IDs.
```

>     a. The subaccounts pool is divided into pieces. For each piece, a query is made to the database to fetch instances.

Reviewer: ```suggestion
>**NOTE:** The subaccounts pool is divided into pieces. For each piece, a query is made to the database to fetch instances.
```

Reviewer: ... or just join it with the previous sentence. There's no sense in making a one-bullet list.

3. Trigger the deprovisioning operation for each instance found in step 2.
Logs inform about the status of each triggered action:

> ```deprovisioning for instance <InstanceID> (SubAccountID: <SubAccountID>) was triggered, operation: <OperationID>```

Reviewer: The code block doesn't render as expected. Try starting ``` with new lines.

Subaccount Cleanup also uses logs to inform about the end of the deprovisioning operation.

## Prerequisites

Subaccount Cleanup requires access to:

> - CIS service to receive all SUBACCOUNT_DELETE events

Reviewer: ```suggestion
- CIS service to receive all `SUBACCOUNT_DELETE` events
```

> - Database to get the instance ID for each subaccount from the SUBACCOUNT_DELETE event

Reviewer: ```suggestion
- Database to get the instance ID for each subaccount ID from the `SUBACCOUNT_DELETE` event
```

---
title: Subaccount Cleanup CronJob
type: Details
---

Each SKR instance in KEB database belongs to global account and subaccount.
Subaccount Cleanup is an application which periodically calls CIS service about SUBACCOUNT_DELETE events 
and based on those events trigger de-provision action on SKR instance which subaccount belongs to.

## Details

All Subaccount Cleanup process is divided into several steps

1. Fetch SUBACCOUNT_DELETE events from CIS service

>     a. CIS client makes first call to CIS service, as a response it gets list of events  divided into pages, 

Reviewer: ```suggestion
    a. CIS client makes a call to the CIS service and as a response, it gets a list of events divided into pages.
    
```

>        total number of events and total number of pages

Reviewer: ```suggestion

```

Reviewer: I'd remove that line as it brings more confusion than explanation. Also, please separate each point with a blank line so that the list renders well.

>     b. CIS client fetches rest of the events by make call to each page one by one

Reviewer: ```suggestion
    b. CIS client fetches the rest of the events by making a call to each page one by one.
    
```

>     c. a subaccount is taken from each event and keep in array

Reviewer: ```suggestion
    c. A subaccount is taken from each event and kept in an array.
    
```

>     d. when CIS client ends, display logs with information how many subaccounts fetched

Reviewer: ```suggestion
    d. When the CIS client ends its workflow, it displays logs with information on how many subaccounts were fetched.
```

> 2. Find all instances in KEB database based on subaccounts fetched in step 1

Reviewer: ```suggestion
2. Find all instances in the KEB database based on the fetched subaccounts.
```

>     a. the whole subaccounts pool is divided into pieces

Reviewer: I wouldn't divide it into separate points. Make a full sentence out of it and add it to point no.2:

```
The subaccounts pool is divided into pieces. For each piece, a query is made to the database to fetch instances.
```

    b. for each pieces query to database is made simultaneously to fetch instances

> 3. Trigger de-provisioning operation for each found instance in step 2

Reviewer: ```suggestion
3. Trigger the deprovisioning operation for each instance found in step 2.
```

>     a. for each instance found, de-provision action is triggered on broker client

Reviewer: I wouldn't use a list here as well. Change the points into full sentences:

```
Logs inform about the status of each triggered action:
<logs>
Subaccount Cleanup also uses logs to inform about the end of the deprovisioning operation.
```

       logs informs for each triggered action:

       ```deprovisioning for instance <InstanceID> (SubAccountID: <SubAccountID>) was triggered, operation: <OperationID>```
    b. at the end, Subaccount Cleanup informs about the end of his work through the appropriate log

## Prerequisites

Subaccount Cleanup requires access to:
- CIS service to receive all SUBACCOUNT_DELETE events

> - Database to get an Instance ID for each Subaccount from SUBACCOUNT_DELETE event

Reviewer: ```suggestion
- Database to get the instance ID for each subaccount from the SUBACCOUNT_DELETE event
```

- Kyma Environment Broker to trigger SKR instance deprovisioning

> ## Configuration

Reviewer: ```suggestion
## Configuration

Use the following environment variables to configure the application:
```

| Environment variable                       | Description                                                                                                                        
|--------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------

> | **APP_CLIENT_VERSION** | Specifies from which version of the service events will be fetched: v1.0 or v2.0

Reviewer: ```suggestion
| **APP_CLIENT_VERSION** | Specifies the service version from which events are fetched. The possible values are  `v1.0` or `v2.0`.
```

Reviewer: Consider adding `Default value` and `Required` columns, if necessary.

Author: All values are required, default values are pointless because there are no default passwords

> | **APP_CIS_CLIENT_ID** | Specifies the client id for the OAuth2 authentication in CIS.

Reviewer: ```suggestion
| **APP_CIS_CLIENT_ID** | Specifies the client ID for the OAuth2 authentication in CIS.
```

> | **APP_CIS_CLIENT_SECRET** | Specifies the client secret for the OAuth2 authentication in CIS.

Reviewer: ```suggestion
| **APP_CIS_CLIENT_SECRET** | Specifies the client Secret for the OAuth2 authentication in CIS.
```

Author: It is not k8s resource Secret, it is secret in meaning of password. IMO it shouldn't be in capital letter.

> | **APP_CIS_AUTH_URL** | Specifies the CIS OAuth token endpoint.

Reviewer: ```suggestion
| **APP_CIS_AUTH_URL** | Specifies the endpoint for the CIS OAuth token.
```

> | **APP_CIS_EVENT_SERVICE_URL** | Specifies the CIS events endpoint.

Reviewer: ```suggestion
| **APP_CIS_EVENT_SERVICE_URL** | Specifies the endpoint for CIS events.
```

| **APP_DATABASE_USER** | Specifies the username for the database. 
| **APP_DATABASE_PASSWORD** | Specifies the user password for the database. 
| **APP_DATABASE_HOST** | Specifies the host of the database. 
| **APP_DATABASE_PORT** | Specifies the port for the database. 
| **APP_DATABASE_NAME** | Specifies the name of the database. 

> | **APP_DATABASE_SSL_MODE** | Activates the SSL mode for PostgrSQL. See all the possible values [here](https://www.postgresql.org/docs/9.1/libpq-ssl.html).  

Reviewer: ```suggestion
| **APP_DATABASE_SSL_MODE** | Activates the SSL mode for PostgreSQL. For reference, see the list of [all the possible values](https://www.postgresql.org/docs/9.1/libpq-ssl.html).  
```

| **APP_BROKER_URL**                             | Specifies the Kyma Environment Broker URL.                                                                                         

> | **APP_BROKER_TOKEN_URL**                       | Specifies the Kyma Environment Broker OAuth token endpoint.                                                                        

Reviewer: ```suggestion
| **APP_BROKER_TOKEN_URL**                       | Specifies the endpoint for Kyma Environment Broker OAuth token.                                                                        
```

| **APP_BROKER_CLIENT_ID**                       | Specifies the username for the OAuth2 authentication in KEB.                                                                       
| **APP_BROKER_CLIENT_SECRET**                   | Specifies the password for the OAuth2 authentication in KEB.                                                                       

> | **APP_BROKER_SCOPE**                           | Specifies the scope for the OAuth2 authentication in KEB.

Reviewer: ```suggestion
| **APP_BROKER_SCOPE**                           | Specifies the scope of the OAuth2 authentication in KEB.
```
